### PR TITLE
[Ldap] Ldap Entry case-sensitive attribute key option

### PR DESCRIPTION
**Pull request by karlshea**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | N/A
| License       | MIT
| Doc PR        | N/A

See PR #36432

### CODE REVIEW

File: src/Symfony/Component/Ldap/Entry.php
Comment (jderusse) [inline src/Symfony/Component/Ldap/Entry.php:25]: should be initialized to an empty array

**Review by jderusse — CHANGES_REQUESTED** [review]
Thank you for this PR.

This is not a bug fix, but a new feature, and should target Symfony 5.x

**Comment by karlshea** [comment]
@jderusse Thanks!

The reason for this PR is for the Drupal LDAP module, and the issue is here: https://www.drupal.org/project/ldap/issues/3126807

That module is using 4.4. Are you saying that this PR could only appear in 5.x? Or that I should target 5.x and it could be backported?

**Comment by derrabus** [comment]
> Are you saying that this PR could only appear in 5.x? 

Yes. 4.4 is closed for features, so you would need to target 5.x.

**Comment by karlshea** [comment]
Just to be clear so I can update the Drupal issue, there is no possibility of this appearing in 4.x?

File: src/Symfony/Component/Ldap/Entry.php
Comment (derrabus) [inline src/Symfony/Component/Ldap/Entry.php]:
```suggestion
    public function hasAttribute($name /* , $caseSensitive = true */)
    {
        $caseSensitive = 2 > \func_num_args() || true === \func_get_arg(1);
```

Adding a parameter to the signature of a public method is a BC break. You can document and use it already, but we have to wait until Symfony 6 to add it to the signature.

File: src/Symfony/Component/Ldap/Entry.php
Comment (derrabus) [inline src/Symfony/Component/Ldap/Entry.php]: Same problem.

File: src/Symfony/Component/Ldap/Entry.php
Comment (derrabus) [inline src/Symfony/Component/Ldap/Entry.php]:
```suggestion
     */
    private function getAttributeKey(string $name, bool $caseSensitive = true): ?string
```

File: src/Symfony/Component/Ldap/Entry.php
Comment (derrabus) [inline src/Symfony/Component/Ldap/Entry.php]:
```suggestion
        return $this->lowerMap[strtolower($name)] ?? null;
```

File: src/Symfony/Component/Ldap/Entry.php
Comment (derrabus) [inline src/Symfony/Component/Ldap/Entry.php]:
```suggestion
        if (null === $attributeKey) {
```

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Ldap/Entry.php
Comment (karlshea) [reply src/Symfony/Component/Ldap/Entry.php]: Is the Symfony code style to comment out the parameter like you have in the suggestion? Or should I remove it entirely?

File: src/Symfony/Component/Ldap/Entry.php
Comment (derrabus) [inline src/Symfony/Component/Ldap/Entry.php]: Please the new parameter as a comment already, as suggested in my previous review. This way, we don't forget to add it in 6.0.

File: src/Symfony/Component/Ldap/Entry.php
Comment (derrabus) [reply src/Symfony/Component/Ldap/Entry.php]: Yes, the comment reminds us to enable the parameter on the next major version.

**Comment by derrabus** [comment]
> Just to be clear so I can update the Drupal issue, there is no possibility of this appearing in 4.x?

There is no 4.x anymore. We're only bugfixing 4.4 and feature development has shifted to 5.x. I cannot make a final decision here, but to me this change does not look like a bugfix. Is it a big problem to support 5.x as well?

**Comment by karlshea** [comment]
> There is no 4.x anymore. We're only bugfixing 4.4 and feature development has shifted to 5.x. I cannot make a final decision here, but to me this change does not look like a bugfix. Is it a big problem to support 5.x as well?

I don't think Drupal 8/Drupal 9 support Symfony 5.x, so that module is stuck with 4.4. They're already using a fork that only contains the Ldap components though, so I'm not sure what direction the maintainer will eventually go.

**Comment by derrabus** [comment]
> I don't think Drupal 8/Drupal 9 support Symfony 5.x, so that module is stuck with 4.4.

The LDAP component is pretty much self-contained. You should be able to upgrade `symfony/ldap` to 5.2 while pinning all other components to 4.4.

**Comment by karlshea** [comment]
Good to know! I'll follow up in the Drupal issue and see what they want to do.

Current test failure seems unrelated to Ldap.

**Review by derrabus — APPROVED** [review]
(for 5.x)

**Review by jderusse — APPROVED** [review]

**Comment by derrabus** [comment]
@karlshea I've force-pushed to your branch in order to resolve conflicts with 5.x and to test your changes with our new LDAP integration test suite.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @karlshea.

**Pull request merged**
